File: backend/app/api/v1/endpoint/celebrations.py
from fastapi import APIRouter, Depends, HTTPException, status, Query
from sqlalchemy.orm import Session
from typing import Optional
from app.db.session import get_db
from app.model.celebration import Celebration
from app.schemas.celebrations import CelebrationCreate, CelebrationUpdate, CelebrationOut
from app.dependencies.auth import get_current_user
from app.model.user import User

router = APIRouter()

@router.get("/", response_model=list[CelebrationOut])
def list_celebrations(
    db: Session = Depends(get_db),
    skip: int = Query(0, ge=0),
    limit: int = Query(100, ge=1, le=100),
    is_active: Optional[bool] = None,
    search: Optional[str] = None,
    current_user: User = Depends(get_current_user),
):
    """
    Listar celebraciones con paginación y filtros opcionales.
    """
    query = db.query(Celebration)

    if is_active is not None:
        query = query.filter(Celebration.is_active == is_active)
    if search:
        query = query.filter(Celebration.title.ilike(f"%{search}%"))

    total = query.count()
    items = query.offset(skip).limit(limit).all()
    return items

@router.post("/", response_model=CelebrationOut, status_code=status.HTTP_201_CREATED)
def create_celebration(
    celebration_data: CelebrationCreate,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    """
    Crear una nueva celebración. Solo admin/editor pueden crear.
    """
    if current_user.role not in ["admin", "editor"]:
        raise HTTPException(
            status_code=status.HTTP_403_FORBIDDEN,
            detail="No tienes permiso para crear celebraciones"
        )

    new_celebration = Celebration(**celebration_data.model_dump())
    db.add(new_celebration)
    db.commit()
    db.refresh(new_celebration)
    return new_celebration

@router.get("/{celebration_id}", response_model=CelebrationOut)
def get_celebration(
    celebration_id: int,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    celebration = db.query(Celebration).filter(Celebration.id == celebration_id).first()
    if not celebration:
        raise HTTPException(status_code=404, detail="Celebración no encontrada")
    return celebration

@router.put("/{celebration_id}", response_model=CelebrationOut)
def update_celebration(
    celebration_id: int,
    celebration_data: CelebrationUpdate,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    """
    Actualizar una celebración. Solo admin/editor pueden actualizar.
    """
    celebration = db.query(Celebration).filter(Celebration.id == celebration_id).first()
    if not celebration:
        raise HTTPException(status_code=404, detail="Celebración no encontrada")

    if current_user.role not in ["admin", "editor"]:
        raise HTTPException(
            status_code=status.HTTP_403_FORBIDDEN,
            detail="No tienes permiso para modificar celebraciones"
        )

    update_data = celebration_data.model_dump(exclude_unset=True)
    for key, value in update_data.items():
        setattr(celebration, key, value)

    db.commit()
    db.refresh(celebration)
    return celebration

@router.delete("/{celebration_id}", status_code=status.HTTP_204_NO_CONTENT)
def delete_celebration(
    celebration_id: int,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    """
    Eliminar una celebración. Solo admin puede eliminar.
    """
    celebration = db.query(Celebration).filter(Celebration.id == celebration_id).first()
    if not celebration:
        raise HTTPException(status_code=404, detail="Celebración no encontrada")

    if current_user.role != "admin":
        raise HTTPException(
            status_code=status.HTTP_403_FORBIDDEN,
            detail="No tienes permiso para eliminar celebraciones"
        )

    db.delete(celebration)
    db.commit()

File: frontend/lib/api/celebrations.ts
import { Celebration, CelebrationCreate, CelebrationUpdate } from '@/src/types/celebration';
import {getApiBaseUrl} from './client';

const API_URL = getApiBaseUrl();

export async function getCelebrations(
  params?: {
    skip?: number;
    limit?: number;
    is_active?: boolean;
    search?: string;
  },
  token?: string
): Promise<Celebration[]> {
  if (!token) throw new Error('No autenticado');

  const query = new URLSearchParams();
  if (params?.skip) query.append('skip', params.skip.toString());
  if (params?.limit) query.append('limit', params.limit.toString());
  if (params?.is_active !== undefined) query.append('is_active', params.is_active.toString());
  if (params?.search) query.append('search', params.search);

  const res = await fetch(`${API_URL}/celebrations?${query.toString()}`, {
    headers: {
      'Authorization': `Bearer ${token}`,
      'Content-Type': 'application/json',
    },
  });

  if (!res.ok) {
    const errorText = await res.text();
    throw new Error(`Error al cargar celebraciones: ${res.status} - ${errorText}`);
  }
  return res.json();
}

export async function getCelebration(id: number, token: string): Promise<Celebration> {
  if (!token) throw new Error('No autenticado');

  const res = await fetch(`${API_URL}/celebrations/${id}`, {
    headers: {
      'Authorization': `Bearer ${token}`,
      'Content-Type': 'application/json',
    },
  });

  if (!res.ok) {
    const errorText = await res.text();
    throw new Error(`Celebración no encontrada: ${res.status} - ${errorText}`);
  }
  return res.json();
}

export async function createCelebration(data: CelebrationCreate): Promise<Celebration> {
  const res = await fetch(`${API_URL}/celebrations`, {
    method: 'POST',
    headers: { 'Content-Type': 'application/json' },
    credentials: 'include',
    body: JSON.stringify(data),
  });

  if (!res.ok) {
    const error = await res.json();
    throw new Error(error.detail || 'Error al crear celebración');
  }
  return res.json();
}

export async function updateCelebration(id: number, data: CelebrationUpdate): Promise<Celebration> {
  const res = await fetch(`${API_URL}/celebrations/${id}`, {
    method: 'PUT',
    headers: { 'Content-Type': 'application/json' },
    credentials: 'include',
    body: JSON.stringify(data),
  });

  if (!res.ok) {
    const error = await res.json();
    throw new Error(error.detail || 'Error al actualizar celebración');
  }
  return res.json();
}

export async function deleteCelebration(id: number): Promise<void> {
  const res = await fetch(`${API_URL}/celebrations/${id}`, {
    method: 'DELETE',
    credentials: 'include',
  });

  if (!res.ok) {
    const error = await res.json();
    throw new Error(error.detail || 'Error al eliminar celebración');
  }
}